HISTORY.md

```diff
@@ -27,6 +27,7 @@
 * Added auto resume function to automatically recover the DB from background Retryable IO Error. When retryable IOError happens during flush and WAL write, the error is mapped to Hard Error and DB will be in read mode. When retryable IO Error happens during compaction, the error will be mapped to Soft Error. DB is still in write/read mode. Autoresume function will create a thread for a DB to call DB->ResumeImpl() to try the recover for Retryable IO Error during flush and WAL write. Compaction will be rescheduled by itself if retryable IO Error happens. Auto resume may also cause other Retryable IO Error during the recovery, so the recovery will fail. Retry the auto resume may solve the issue, so we use max_bgerror_resume_count to decide how many resume cycles will be tried in total. If it is <=0, auto resume retryable IO Error is disabled. Default is INT_MAX, which will lead to a infinit auto resume. bgerror_resume_retry_interval decides the time interval between two auto resumes.
+* A new option `std::shared_ptr<FileChecksumGenFactory> file_checksum_gen_factory` is added to `BackupableDBOptions`. The default value for this option is `nullptr`. If this option is null, the default backup engine checksum function (crc32c) will be used for creating, verifying, or restoring backups. If it is not null and is set to the DB custom checksum factory, the custom checksum function used in DB will also be used for creating, verifying, or restoring backups, in addition to the default checksum function (crc32c). If it is not null and is set to a custom checksum factory different than the DB custom checksum factory (which may be null), BackupEngine will use only the default checksum function (crc32c) for creating, verifying, or restoring backups.
```

Comment:
Now that I see we are supporting different checksum functions from same factory, I'm more concerned about the case of users backup checksum factory quietly not supporting the functions used in the DB. Proposal: it should be an error (invalid argument) if BackupableDBOptions::file_checksum_gen_factory is set and we try to back up a DB with custom checksum not supported by it. If someone needs to backup a DB for which they don't have access to (at least one) custom checksum, I think the best remediation is to require them to unset file_checksum_gen_factory on backup, to acknowledge degraded data integrity checking, rather than very quietly degrading data integrity checking.

Reply:
Right, I actually implemented this in an earlier draft but with different error status (`NotSupported()` for a different factory and `NotFound()` for a different generator). The code is still there but the error status is suppressed. So it would be relatively easy to do it in this PR or later :)